rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java

```diff
@@ -99,35 +99,48 @@ public Long call()
           propagateCancel(this.getClass().getName());
         }
         log.debug(String.format("Executing ReadRequest: [%d, %d, %d, %d, %d]", readRequest.getBackendReadStart(), readRequest.getBackendReadEnd(), readRequest.getActualReadStart(), readRequest.getActualReadEnd(), readRequest.getDestBufferOffset()));
-        inputStream.seek(readRequest.backendReadStart);
-
         int prefixBufferLength = (int) (readRequest.getActualReadStart() - readRequest.getBackendReadStart());
         int suffixBufferLength = (int) (readRequest.getBackendReadEnd() - readRequest.getActualReadEnd());
         log.debug(String.format("PrefixLength: %d SuffixLength: %d", prefixBufferLength, suffixBufferLength));
 
+        // Possible only for first readRequest in chain
         if (prefixBufferLength > 0) {
+          // Streaming read because next read is guaranteed to be contiguous
+          inputStream.seek(readRequest.backendReadStart);
           log.debug(String.format("Trying to Read %d bytes into prefix buffer", prefixBufferLength));
           totalPrefixRead += readIntoBuffer(affixBuffer, 0, prefixBufferLength, inputStream);
-          log.debug(String.format("Read %d bytes into prefix buffer", prefixBufferLength));
-          copyIntoCache(fileChannel, directBuffer, affixBuffer, 0, prefixBufferLength, readRequest.backendReadStart);
-          log.debug(String.format("Copied %d prefix bytes into cache", prefixBufferLength));
+          int written = copyIntoCache(fileChannel, directBuffer, affixBuffer, 0, prefixBufferLength, readRequest.backendReadStart);
+          log.debug(String.format("Copied %d prefix bytes into cache", written));
         }
 
         log.debug(String.format("Trying to Read %d bytes into destination buffer", readRequest.getActualReadLengthIntUnsafe()));
-        int readBytes = readIntoBuffer(readRequest.getDestBuffer(), readRequest.destBufferOffset, readRequest.getActualReadLengthIntUnsafe(), inputStream);
-        log.debug(String.format("Read %d bytes into destination buffer", readBytes));
-        copyIntoCache(fileChannel, directBuffer, readRequest.destBuffer, readRequest.destBufferOffset, readBytes, readRequest.actualReadStart);
-        log.debug(String.format("Copied %d requested bytes into cache", readBytes));
-        totalRequestedRead += readBytes;
+        int readBytes;
+        if (prefixBufferLength > 0 || suffixBufferLength > 0) {
+          // For single readRequest in chain, prefix and suffix both may be present
+          // seek needed in case of just suffix being present, otherwise it is no-op
+          inputStream.seek(readRequest.actualReadStart);
+          readBytes = readIntoBuffer(readRequest.getDestBuffer(), readRequest.destBufferOffset, readRequest.getActualReadLengthIntUnsafe(), inputStream);
```

Comment:
checkState(readBytes == readRequest.getActualReadLengthIntUnsafe())

Reply:
This check is currently there in `if (suffixBufferLength > 0)` block, I can move that here. Earlier we could see exception due to this in `Bookeeper#readDataInternal`, with the changes there we can add validation here

Reply:
Ya, lets move it here for readability. Also lets add checkState of each of the reads here: prefix, data and suffix as we are marking block in cache later even if we didnt read all data.